Blake3.lean

```diff
@@ -53,88 +58,105 @@ opaque initDeriveKey (context : @& ByteArray) : Hasher
-
-theorem size_of_extract : (ByteArray.extract x b e).size = e - b := sorry
-
-namespace StatefulHashObject
-  -- TODO: Can we avoid gaving explicit 'hash' field that includes key in the beginning
-  structure Sponge where
-    hasher: Hasher
-    hash : ByteArray
-
-  -- TODO: We have to enforce label to contain: 1) software identifier, 2) timestamp, 3) intended use-case 4) version
-  -- This probably can be implemented using some RegExp
-  -- Good label example: 'ix 2025-01-01 16:18:03 content-addressing v1'
-  def checkLabel (label : String) : String := label
-
-  def Sponge.new (checkLabel: String → String) (label : String) : Sponge := {
-    hasher := Hasher.initDeriveKey (String.toUTF8 (checkLabel label))
-    hash := ⟨#[]⟩
-  }
-
-  -- TODO: Ideally we should have automatic re-keying of the hasher, based on some counter
-  -- (like in Rust reference: https://github.com/storojs72/BLAKE3/blob/sho/reference_impl/reference_impl.rs#L467)
-  def Sponge.absorb (sponge: Sponge) (input: ByteArray) : Sponge := {
-    sponge with hasher := sponge.hasher.update input
-  }
-
-  def extractOutput (hash: ByteArray) : ByteArray := ByteArray.extract hash (2 * BLAKE3_KEY_LEN) (hash.size + 2 * BLAKE3_KEY_LEN)
-  def extractKey (hash: ByteArray) : Blake3Key := Blake3Key.ofBytes (ByteArray.extract hash 0 BLAKE3_KEY_LEN) size_of_extract
-
-  -- returns tuple
-  def Sponge.squeeze (sponge: Sponge) (length: Nat) : Sponge × ByteArray :=
-  let tmp := sponge.hasher.finalize (USize.ofNat (length + 2 * BLAKE3_KEY_LEN))
-  let newSponge := {
-    hasher :=  Hasher.initKeyed (extractKey tmp)
-    hash := extractOutput tmp
-  }
-  (newSponge, newSponge.hash)
-
-
-  --def extract_output (sponge: Sponge) : ByteArray := ByteArray.extract sponge.hash (2 * BLAKE3_KEY_LEN) (sponge.hash.size + 2 * BLAKE3_KEY_LEN)
-  --def extract_key (hash: ByteArray) : ByteArray := ByteArray.extract hash 0 BLAKE3_KEY_LEN
-
-  -- This has to be "periodically" invoked manually for security reasons
-  def Sponge.ratchet (sponge: Sponge): Sponge := {
-    hasher := Hasher.initKeyed (extractKey sponge.hash)
-    hash := ⟨#[]⟩
-  }
-
-end StatefulHashObject
+  hasher.finalizeBlake3Hash
+
+structure Sponge where
+  hasher : Hasher
+  counter : Nat
+
+namespace Sponge
+
+abbrev ABSORB_MAX_BYTES := UInt32.size - 1
+abbrev DEFAULT_REKEYING_STAGE := UInt16.size - 1
+
+def init (label : String) (_h : ¬label.isEmpty := by decide) : Sponge :=
+  ⟨Hasher.initDeriveKey label.toUTF8, 0⟩
+
+def ratchet (sponge : Sponge) : Sponge :=
+  let key := sponge.hasher.finalizeBlake3Key
+  { sponge with hasher := Hasher.initKeyed key }
+
+def absorb (sponge : Sponge) (bytes : ByteArray)
+    (_h : bytes.size < ABSORB_MAX_BYTES := by norm_cast) : Sponge :=
+  let highCounter := sponge.counter >= DEFAULT_REKEYING_STAGE
+  let sponge := if highCounter then sponge.ratchet else sponge
+  -- Is `ratchet` supposed to reset `counter` to 0? Or should it be reset
+  -- everytime `highCounter` is `true`?
```

Comment:
What's supposed to happen here? If we don't do anything, `highCounter` will always be `true` after the first time it's `true`.